https: disallow boolean types for `key` and `cert` options

yjimk opened the pull request:
*This is my first PR on the Node repository, your feedback is greatly appreciated.*

When using https.createServer, passing boolean values for `key` and `cert` properties in the options object parameter doesn't throw an error an could lead to potential issues if they're accidentally passed.

This PR aims to throw a reasonable error if a boolean was passed to either of those properties.

Fixes: https://github.com/nodejs/node/issues/12802

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
https

addaleax commented on lib/https.js:
```diff
@@ -40,6 +40,12 @@ function Server(opts, requestListener) {
+    throw new Error('"options.key" must not be a boolean');
+
+  if (opts && typeof opts.cert === 'boolean')
+    throw new Error('"options.cert" must not be a boolean');
```
By the way, we try to migrate to a [new error system](https://github.com/nodejs/node/blob/master/doc/guides/using-internal-errors.md) that allows for more flexibility in the future, it would be cool if you think you could take a look at using that.

addaleax commented on test/parallel/test-https-options-boolean-check.js:
```diff
@@ -0,0 +1,84 @@
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
```
We don’t need the copyright header for new files.

refack commented:
Hello @yjimk and welcome. Thank you very much for you contribution 🥇.
You've chosen an interesting case... While your change is correct, it checks for a very specific scenario so I'll give you a few of my observations:
1. I just followed the the [documentation](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options) and found that some of the options passed to `https.createServer` are than passed to `tls.createSecureContext`, so it's probably a better place to put the validation. https://github.com/nodejs/node/blob/262d4b29cba3650ac6d826149e1a7e5e2a2a087f/lib/_tls_common.js#L84-L107
2. The documentation state that `cert` and `key` can be  `<string> | <string[]> | <Buffer> | <Buffer[]> | <Object[]>` so it is probably better to assert that the option _is_ one of those type, rather than trying to exclude everything else.
3. We are migration from using simply `throw new Error()`  to a more structure error mechanism that includes an error code like - https://github.com/nodejs/node/blob/262d4b29cba3650ac6d826149e1a7e5e2a2a087f/lib/path.js#L28
the full guide is in https://github.com/nodejs/node/blob/master/doc/guides/using-internal-errors.md

Trott commented on lib/https.js:
```diff
@@ -40,6 +40,12 @@ function Server(opts, requestListener) {
   }
   opts = util._extend({}, opts);
 
+  if (opts && typeof opts.key === 'boolean')
```
Line 41 guarantee that `opts` is an object so there is no need to check `opts` here or in line 46.

Trott replied:
I don't think checking against all invalid values is the way to go. I think it should confirm it is a valid value and reject everything else.

Trott commented on test/parallel/test-https-options-boolean-check.js:
```diff
@@ -0,0 +1,84 @@
+  https.createServer({
+    key: true,
+    cert: fs.readFileSync(`${common.fixturesDir}/keys/agent1-cert.pem`)
+  }), /"options\.key" must not be a boolean/);
```
These regular expressions would be better if they matched the entire message (that is, if they started with `^` and ended with `$`).

Trott commented on lib/https.js:
```diff
@@ -40,6 +40,12 @@ function Server(opts, requestListener) {
   opts = util._extend({}, opts);
 
+  if (opts && typeof opts.key === 'boolean')
+    throw new Error('"options.key" must not be a boolean');
```
This and the other `Error` should be a `TypeError`.

cjihrig commented on test/parallel/test-https-options-boolean-check.js:
```diff
@@ -0,0 +1,84 @@
+    cert: fs.readFileSync(`${common.fixturesDir}/keys/agent1-cert.pem`)
+  }));
+
+assert.throws(() =>
```
You can reduce code duplication a lot here. All of these `assert.throws()` calls are essentially the same, with only the value of `key`, `cert`, and the regular expression changing. You can do something like this instead:

```javascript
const key = fs.readFileSync(`${common.fixturesDir}/keys/agent1-key.pem`);
const cert = fs.readFileSync(`${common.fixturesDir}/keys/agent1-cert.pem`);
const invalidKeyRE = /^"options\.key" must not be a boolean$/;
const invalidCertRE = /^"options\.cert" must not be a boolean$/;

[
  [true, cert, invalidKeyRE],
  // Fill in other cases here
].forEach((params) => {
  assert.throws(() => {
    https.createServer({
      key: params[0],
      cert: params[1]
    });
  }, params[2]);
});
```

silverwind commented:
I agree with @Trott, better verify for validity instead of checking single invalid values like booleans.

Also, I think this should also apply to `tls.createServer` and `tls.createSecureContext`, so I'd say the change should be done in `tls.createSecureContext` instead of only in `https`.

yjimk commented:
Thank you all for the super helpful and detailed feedback! I'll make those changes and see if I can get closer.

yjimk commented:
I believe I might have addressed all of the comments left with my latest commit. The new code is a bit of a departure from the old, so I hope it is OK and would appreciate any tips.

Particularly worried about using that function `validateKeyCert` from a naming perspective and also whether it conforms to the right method for type checking.

There are a few other things in that file that I noticed are a different style compared to other parts of the code, such as usage of `for` loops and `var` over `let` and `const`.

Thanks again for your great feedback and comments, it's been a fantastic learning experience.

addaleax commented on lib/_tls_common.js:
```diff
@@ -52,6 +54,13 @@ function SecureContext(secureProtocol, secureOptions, context) {
 }
 
+function validateKeyCert(value, type) {
+  if (typeof value !== 'string' && !(value instanceof Buffer))
```
I think this is too strict, and `instanceof Buffer` is not a foolproof way to check for Buffers. Can you use `ArrayBuffer.isView(value)` instead and adjust the message accordingly (Buffers are just special Uint8Arrays, but all Uint8Arrays and other ArrayBufferViews should work, I think.).

yjimk replied:
Ah, brilliant! Replacing the two check works perfectly. I'll change this over and add some additional tests.

Regarding the new error message, it is currently `['string', 'buffer']` and I'm thinking it should change to `['string', 'ArrayBuffer']` with perhaps `string` => `String`. Would you have an opinion on how that should look?

addaleax replied:
I don’t think we can list `ArrayBuffer`, as those are not ArrayBufferViews themselves.

Maybe `['string', 'Buffer', 'TypedArray', 'DataView']`? That would match the message here: https://github.com/nodejs/node/blob/cde272a066923a1b93a10a8121b0ce572308d13c/lib/zlib.js#L231

yjimk replied:
Perfect! Thank you, I'll create a commit for this change.

jasnell commented on test/parallel/test-https-options-boolean-check.js:
```diff
@@ -0,0 +1,126 @@
+const https = require('https');
+const fs = require('fs');
+
+const keyBuff = fs.readFileSync(`${common.fixturesDir}/keys/agent1-key.pem`);
```
There's been some recent work to simplify fixtures access...

example...
```js
const fixtures = require('../common/fixtures');

const keyBuff = fixtures.readKey('agent1-key.pem');
```

yjimk replied:
Ah OK, cool. Thanks, I'll change that over too. The original test file was copied over from another and amended, hence the previous inclusion of the license comments and usage of the `fs.readFileSync` method of including the pem files. Is there are issue/task that you're aware of to update the older tests to use the newer methods? Something like that I feel like would be a useful task for me to take up in future.

fl0w commented:
*Since this PR relates to #12802 which is marked as "good first contribution", I'm going to ask a question (as I also intended on grabbing this). If this is inappropriate, delete my question and I'll take it to IRC or SO*

From e92e64ad extended commit message:
> Arrays are now iterated using `map` instead of `for` for affected sections

What's the benefit of changing the loop type? I can't seem to find related review comment that (probably) requested the change.

yjimk commented:
RE: @fl0w, mostly due to personal preference toward `map` and other functional programming methods rather than `for` or `forEach` loops.

This tweet (although it mentions `forEach` instead of `for`) is the basis for my way of thinking.

https://twitter.com/_ericelliott/status/710571765844615168

I'd love to hear the `for`s and againsts on it though

...sorry

yjimk commented:
I've changed the type checking for buffer, but my knowledge of ArrayBuffers and DataViews is pretty limited.

silverwind commented on lib/_tls_common.js:
```diff
@@ -71,10 +80,12 @@ exports.createSecureContext = function createSecureContext(options, context) {
+      options.ca.map((ca) => {
+        validateKeyCert(ca, 'ca');
+        c.context.addCACert(ca);
+      });
```
This `map` can be changed to a `forEach` because you're not using the return value.

silverwind commented on lib/_tls_common.js:
```diff
@@ -83,9 +94,12 @@ exports.createSecureContext = function createSecureContext(options, context) {
+      options.cert.map((cert) => {
+        validateKeyCert(cert, 'cert');
+        c.context.setCert(cert);
+      });
```
Same here.

silverwind commented on lib/_tls_common.js:
```diff
@@ -96,12 +110,12 @@ exports.createSecureContext = function createSecureContext(options, context) {
+      options.key.map((k) => {
+        validateKeyCert(k.pem || k, 'key');
+        c.context.setKey(k.pem || k, k.passphrase || options.passphrase);
+      });
```
And here.

silverwind commented:
> RE: @fl0w, mostly due to personal preference toward map and other functional programming methods rather than for or forEach loops.

Sorry, just saw your comment after review. I personally see no benefit of using `map` over `forEach` if you're not going to return something/use the return value.

Node.js uses `for` a lot because as far as we know, it is still considerably faster than the prototype methods, thought that might have changed with TurboFan.

yjimk commented:
No problem, I'll change it over to use the forEach method. :)

jasnell commented:
> thought that might have changed with TurboFan

It does not appear to have changed by much.

BridgeAR commented:
Landed in a7dccd040d72ce7de61d9160ec031420c52a49d4
